### ISSUE TITLE
SHA-1 prefix spec collision when specifying hashes for dependencies

### ISSUE THREAD

**Issue by wscullin** [spack/spack#3717]
I'm shortening flags and compiler options for the sake of brevity as the total length is 1134 chars, but can provide full paths if it matters. This may also be a variant of #1559, but it seems to be a slightly different animal.

Using the Intel compilers, pixman will only build and pass tests when optimization is set to -O0. It depends on libpng which in turn depends on zlib. I don't want to build everything -O0, so I have dependencies prebuilt with flags that generally work on our system:

```bash
[derp]$ spack find -fl %intel
==> 3 installed packages.
-- cray-cnl6-mic_knl / intel@17.0.2.174.4.9.4.0.0.0.7.5.3 -------
lggn6f6 libpng@1.6.27%intel cflags="-mtune=knl -g -fp-model:source ..."

q6jce7a pkg-config@0.29.1%intel cflags="-mtune=knl -g -fp-model:source ..."

lfr2nzr zlib@1.2.11%intel cflags="-mtune=knl -g -fp-model:source ..."

[derp]$
```

I then go to build pixman and spack sees a collision with libpng's zlib dependency:
```bash
[derp]$ spack install pixman cflags="-O0" %intel@17.0.2.174.4.9.4.0.0.0.7.5.3 ^/lggn6f6
==> Installing pixman
==> Installing libpng
==> Error: Specs zlib@1.2.11%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -O0 -fp-model:source ..." +pic+shared arch=cray-cnl6-mic_knl and zlib@1.2.11%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -fp-model:source ..." +pic+shared arch=cray-cnl6-mic_knl have the same SHA-1 prefix!
[derp]$
```

Looking at the concrete spec for what I asked to install, it looks like an issue in the concretized spec:
```bash
[derp]$ spack spec -L pixman ^/lggn6f6 cflags="-O0"
Input spec
--------------------------------
2o2mp55isbdd7vxariybt3apo2kaomrs  pixman cflags="-O0" 
lggn6f6bhf7dm5w3ttosr27jqpj7ojsn      ^libpng@1.6.27%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -fp-model:source ..."  arch=cray-cnl6-mic_knl 
lfr2nzrs5673spgkn4y7me2ohq5rw5oh          ^zlib@1.2.11%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -fp-model:source ..." +pic+shared arch=cray-cnl6-mic_knl 

Normalized
--------------------------------
z566qe2kizqbln7ezlw5b3vwdccsr4gc  pixman cflags="-O0" 
lggn6f6bhf7dm5w3ttosr27jqpj7ojsn      ^libpng@1.6.27%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -fp-model:source ..."  arch=cray-cnl6-mic_knl 
lfr2nzrs5673spgkn4y7me2ohq5rw5oh          ^zlib@1.2.11%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -fp-model:source ..." +pic+shared arch=cray-cnl6-mic_knl 
tiolp4xnoepfywxfos3ygzir7r3xut22      ^pkg-config

Concretized
--------------------------------
wutzdizhy74xm6rmudjre26rclzy52bw  pixman@0.34.0%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -O0 -fp-model:source ..."  arch=cray-cnl6-mic_knl 
lggn6f6bhf7dm5w3ttosr27jqpj7ojsn      ^libpng@1.6.27%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -O0 -fp-model:source ..." arch=cray-cnl6-mic_knl 
lfr2nzrs5673spgkn4y7me2ohq5rw5oh          ^zlib@1.2.11%intel@17.0.2.174.4.9.4.0.0.0.7.5.3 cflags="-mtune=knl -g -O0 -fp-model:source ..." +pic+shared arch=cray-cnl6-mic_knl 
6dkvb4htwlxxc57iymlt6uin2zjtk6tf      ^pkg-config@0.29.2%intel@17.0.2.174.4.9
```

**Comment by tgamblin** [comment]
I was able to reproduce this.

@becker33: the issue is that the flags logic is pushing compiler flags down into the `libpng/lggn6f6` part of the spec that's already concrete.  It shouldn't set flags there; it should keep already-concrete specs exactly the same.

**Comment by wscullin** [comment]
I have sinned - I'll have the PR tested later today; I thought I'd lost platform concretization with the PR, but I think it might be another change in our environment.

**Comment by tgamblin** [comment]
@wscullin: ping

**Comment by alalazo** [comment]
@wscullin Any news on this issue?

**Comment by alalazo** [comment]
Closing as stale. Should be fixed by #3723
